shared/chat/conversation/list-area/normal/index.desktop.js

```diff
@@ -165,43 +160,251 @@ class Thread extends React.Component<Props, State> {
+        key="topWaypoint"
+        onPositionChange={this._positionChangeTop}
+        conversationIDKey={this.props.conversationIDKey}
+      />
+    )
+
+    const numOrdinals = this.props.messageOrdinals.size
+    let indicies = []
+    for (var idx = 0; idx < numOrdinals; ++idx) {
+      const needNextWaypoint = idx % ordinalsInAWaypoint === 0
+      const isLastItem = idx === numOrdinals - 1
+      if (needNextWaypoint || isLastItem) {
+        if (isLastItem) {
+          indicies.push(idx)
+        }
+
+        if (indicies.length) {
+          const ordinal = this.props.messageOrdinals.get(indicies[0] || 0)
+          if (!ordinal) {
+            throw new Error('Should be impossible')
+          }
+          const key = String(Types.ordinalToNumber(ordinal))
+          waypoints.push(
+            <OrdinalWaypoint key={key} id={key} rowRenderer={this._rowRenderer} indicies={indicies} />
+          )
+          indicies = []
+        }
+      }
+      indicies.push(idx)
+    }
+
+    waypoints.push(
+      <BottomWaypoint
+        key="bottomWaypoint"
+        onPositionChange={this._positionChangeBottom}
+        conversationIDKey={this.props.conversationIDKey}
+      />
+    )
+
+    return waypoints
   }
 
   render() {
-    const rowCount = this._getItemCount()
-    const scrollToIndex = this.state.isLockedToBottom ? rowCount - 1 : undefined
+    const waypoints = this._makeWaypoints()
 
     return (
       <ErrorBoundary>
         <div style={containerStyle} onClick={this._handleListClick} onCopyCapture={this._onCopyCapture}>
           <style>{realCSS}</style>
-          <Virtualized.AutoSizer onResize={this._onResize}>
-            {({height, width}) => (
-              <Virtualized.List
-                conversationIDKey={this.props.conversationIDKey}
-                columnWidth={width}
-                deferredMeasurementCache={this._cellCache}
-                height={height}
-                onScroll={this._onScroll}
-                ref={this._setListRef}
-                rowCount={rowCount}
-                rowHeight={this._cellCache.rowHeight}
-                rowRenderer={this._rowRenderer}
-                scrollToAlignment="start"
-                scrollToIndex={scrollToIndex}
-                style={listStyle}
-                width={width}
-              />
-            )}
-          </Virtualized.AutoSizer>
+          <div style={listStyle} ref={this._listRef} onScroll={this._onScroll}>
+            <div style={this.state.isScrolling ? innerListStyleScrolling : null}>{waypoints}</div>
+          </div>
         </div>
       </ErrorBoundary>
     )
   }
 }
 
+// All the waypoints keep a key to re-render if we get a measure callback
+type TopBottomWaypointProps = {
+  conversationIDKey: Types.ConversationIDKey,
+  onPositionChange: () => void,
+}
+type TopBottomWaypointState = {
+  keyCount: number,
+}
+
+class TopWaypoint extends React.PureComponent<TopBottomWaypointProps, TopBottomWaypointState> {
```

Comment:
can this and `BottomWaypoint` be consolidated and take a prop to tell whether it's top / bottom?